pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py

```diff
@@ -0,0 +1,452 @@
+        'Enter next test, or \'enter\' to see menu',
+    ]
+
+    def __init__(self, dut: Union[Dut, List[Dut]], **kwargs) -> None:  # type: ignore
+        if isinstance(dut, List):
+            for item in dut:
+                if isinstance(item, Dut):
+                    self.dut = item
+                    break
+        else:
+            self.dut = dut
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+        if 'test_menu' not in kwargs:
+            self.get_test_menu()
+
+    def get_test_menu(self) -> None:
+        """
+        Get the test menu of this test app
+
+        Notes:
+            It will do a hard reset after getting the test menu to ensure
+            the patterns that indicate the case is ready not taken by the parser.
+            Please use this function to get the test menu while using this script
+        """
+        self.dut.write('')
+        self.test_menu = self.dut.parse_test_menu()
+        self.dut.serial.hard_reset()
+
+
+class NormalCaseTester(BaseTester):
+    """
+    Tester of normal type case
+
+    Attributes:
+        dut (Dut): Object of the Device under test
+        test_menu (List[UnittestMenuCase]): The list of the cases
+        retry_count (int): number of retries for a single case when it is failed to start
+    """
+
+    def run_all_normal_cases(
+        self,
+        reset: bool = False,
+        timeout: int = DEFAULT_TIMEOUT,
+        retry_count: int = DEFAULT_RETRY_COUNT,
+    ) -> None:
+        """
+        Run all normal cases
+
+        Args:
+            reset: whether do a hardware reset before running the case
+            timeout: timeout in second
+            retry_count (int): number of retries for a single case when it is failed to start
+        """
+        for case in self.test_menu:
+            self.run_normal_case(case, reset, timeout=timeout, retry_count=retry_count)
+
+    def run_normal_case(
+        self,
+        case: UnittestMenuCase,
+        reset: bool = False,
+        timeout: int = DEFAULT_TIMEOUT,
+        retry_count: int = DEFAULT_RETRY_COUNT,
+    ) -> None:
+        """
+        Run a specific normal case
+
+        Notes:
+            Will skip if the case type is not normal
+
+        Args:
+            case: the specific case that parsed in test menu
+            reset: whether do a hardware reset before running the case
+            timeout: timeout in second
+            retry_count (int): number of retries for a single case when it is failed to start
+        """
+        if case.type == 'normal':
+            if reset:
+                self.dut.serial.hard_reset()
+            self.dut.expect(self.ready_pattern_list, timeout=timeout)
+            for retry in range(retry_count):
+                try:
+                    self.dut.write(str(case.index))
+                    self.dut.expect(f'Running {case.name}...', timeout=1)
+                    self.dut.expect_unity_test_output(timeout=timeout)
```

Comment:
@igrr Is your idea to only apply the retry logic at the start of the test ? But when we are checking the unity test output we shouldn't retry . Is that what you meant ?

Reply:
When I tested the script in IDF, I found writing case index to dut might fail  after reset. I guess the failure mainly caused by the hurry writting before the dut get totally ready to receive something, because some delays will help. However, I can't set an exact delay since the required delay are different among different environment, especially on CI.  Therefore, I retired when failed to expect case running pattern.

As for the `expect_unity_test_output` I agree that it shouldn't be included in the `retry` block, because after expecting the `case running` pattern, the case is supposed to be running, when the timeout happened then, it should be treated as a failure.
cc @igrr @hfudev

Reply:
the original idea of the `retry` is for those flaky unit test cases. it should be parsed from the test case's full name. like `TEST_CASE(...[retry=3][timeout=120])`